File: problems/CSES/Sorting_and_Searching/towers/towers.py
import sys
from bisect import bisect_right
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    n = int(input())
    cubes = list(map(int, input().split()))

    tops = []
    for cube in cubes:
        pos = bisect_right(tops, cube)

        if pos == len(tops):
            tops.append(cube)
        else:
            tops[pos] = cube

    print(len(tops))

main()

File: problems/CSES/Sorting_and_Searching/towers/towers.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    vector<ll> cubes(n);
    for (int i = 0; i < n; i++) cin >> cubes[i];

    vector<ll> tops;
    for (ll cube : cubes) {
        size_t pos = upper_bound(tops.begin(), tops.end(), cube) - tops.begin();

        if (pos == tops.size()) {
            tops.push_back(cube);
        } else {
            tops[pos] = cube;
        }
    }

    cout << tops.size() << "\n";

    return 0;
}
